apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/MSALoginActivity.java

```diff
@@ -0,0 +1,396 @@
+        Toast.makeText(this, R.string.sign_out_failed, Toast.LENGTH_SHORT).show();
+        finish();
+    }
+
+    /**
+     * Check from mWebView view if sign in completed and process completion.
+     */
+    private void checkSignInCompletion(String url) {
+        if (url.startsWith(REDIRECT_URL)) {
+            Uri uri = Uri.parse(url);
+            if (mAuthType == AuthenticationProvider.Type.MSA_COMPACT) {
+                String fragment = uri.getFragment();
+                if (fragment != null) {
+
+                    /* Convert fragment to query string to process response. */
+                    checkCompactSignInCompletion(Uri.parse(REDIRECT_URL + "?" + fragment));
+                }
+            } else {
+                checkDelegateSignInCompletion(uri);
+            }
+        }
+    }
+
+    private void checkCompactSignInCompletion(Uri uri) {
+        mRefreshToken = uri.getQueryParameter(REFRESH_TOKEN);
+        mRefreshTokenScope = SCOPES_COMPACT[0];
+        if (!TextUtils.isEmpty(mRefreshToken)) {
+            registerAppCenterAuthentication(uri.getQueryParameter(USER_ID));
+        } else {
+            failSignIn(0, uri.getQueryParameter("error_description"));
+        }
+    }
+
+    private void checkDelegateSignInCompletion(Uri uri) {
+        String code = uri.getQueryParameter("code");
+        if (TextUtils.isEmpty(code)) {
+            failSignIn(0, "error=" + uri.getQueryParameter("error"));
+        } else {
+            getToken(code);
+        }
+    }
+
+    private void registerAppCenterAuthentication(String userId) {
+        AuthenticationProvider.TokenProvider tokenProvider = new AuthenticationProvider.TokenProvider() {
+
+            @Override
+            public void acquireToken(String ticketKey, AuthenticationProvider.AuthenticationCallback callback) {
+
+                /* Refresh token, doing that even on first time to test the refresh code without having to wait 1 hour. */
+                refreshToken(callback);
+            }
+        };
+        AuthenticationProvider provider = new AuthenticationProvider(mAuthType, userId, tokenProvider);
+        AnalyticsTransmissionTarget.addAuthenticationProvider(provider);
+        finish();
+        Toast.makeText(this, R.string.signed_in, Toast.LENGTH_SHORT).show();
+    }
+
+    /**
+     * Get initial access token.
+     */
+    private void getToken(final String code) {
+        Map<String, String> headers = new HashMap<>();
+        headers.put(DefaultHttpClient.CONTENT_TYPE_KEY, "application/x-www-form-urlencoded");
+        sHttpClient.callAsync(TOKEN_URL,
+                DefaultHttpClient.METHOD_POST,
+                headers,
+                new HttpClient.CallTemplate() {
+
+                    @Override
+                    public String buildRequestBody() {
+                        return REDIRECT_URI_PARAM +
+                                CLIENT_ID_PARAM +
+                                "&grant_type=authorization_code" +
```

Comment:
const?

Reply:
Thanks for the feedback but we can clean it up in another task. This is just copied from another place and we don't change any lines of code here.